Article: Dask cluster overview (#88)
jacobtomlinson/website: content/posts/2020/2020-02-19-dask-clusters-overview.md

It is also common in this paradigm for multiple users to share this single cluster, however this is not recommended as the Dask scheduler does not manage users or clients separately and work will be executed on a first come first served bases. Therefore we recommend that users use a cluster one at a time.

### Ephemeral clusters

An ephemeral cluster is one which only exists for the duration of the work. In this case a user may SSH onto the machines, run the commands to set up the cluster, connect a client and perform work, then disconnect and exit the Dask processes. A basic way of doing this would be to create a bash script which calls `ssh` and sets up the cluster. You would run this script in the background while performing your work and then kill it once you are done. We will cover other implementations of this in the coming sections.

Ephemeral clusters allow you to leverage a bunch of machines but free them up again when you are done. This is especially useful when you are using a system like a cloud service or a batch scheduler where you have limited credits, or are charged for provisioned resources.

## Adaptivity

> Ephemeral clusters are also generally easier to scale as you will likely have an automated mechanism for starting workers. The Dask scheduler maintains an estimate of how long it expects the outstanding work to take to complete. If the scheduler has a mechanism for starting and stopping workers then it will scale up the workers in an attempt to complete all outstanding work within 5 seconds. This is referred to as adaptive mode.

Reviewer: ```suggestion
Ephemeral clusters are also generally easier to scale as you will likely have an automated mechanism for starting workers. The Dask scheduler maintains an estimate of how long it expects the outstanding work will take to complete. If the scheduler has a mechanism for starting and stopping workers then it will scale up the workers in an attempt to complete all outstanding work within 5 seconds. This is referred to as adaptive mode.
```

The mechanisms for starting and stopping workers are added via plugins. Many of the implementations we are about to discuss include this logic.

## Connectivity

> Dask uses TCP to communicate between client, scheduler and workers by default. This means that all of these components must be on a TCP/IP network with open routes between the machines. Many connectivity problems step from firewalls or private networks blocking connections between certain components. An example of this would be running Dask on a cloud platform like AWS, but running the Python session and client on your laptop while using the free wifi in a coffee shop. You must ensure you are able to route traffic between components, either by exposing the Dask cluster to the internet or by connecting your laptop to the private network via a VPN or tunnel.

Reviewer 2: Would it be useful here to specify the *direction* of connectivity - from workers to scheduler only, rather than scheduler to workers (I think, at least!). It might be too much detail for a blog post though.

Author: Yeah that's a really good point! Although am I right in thinking that connectivity happens in both directions?

- Worker connects to the scheduler at scheduler address and provides info about worker
- Scheduler then opens an RPC connection back to the worker on the address provided

Reviewer 2: Yeah, writing this comment made me realize I don't actually know! :D

Reviewer 3: Connectivity happens in both directions

> Dask uses TCP to communicate between client, scheduler and workers by default. This means that all of these components must be on a TCP/IP network with open routes between the machines. Many connectivity problems step from firewalls or private networks blocking connections between certain components. An example of this would be running Dask on a cloud platform like AWS, but running the Python session and client on your laptop while using the free wifi in a coffee shop. You must ensure you are able to route traffic between components, either by exposing the Dask cluster to the internet or by connecting your laptop to the private network via a VPN or tunnel.

Reviewer: "but running the Python session and client on your laptop while using the free wifi in a coffee shop."

Seems like something is missing in that sentence above.

Author: ```suggestion
Dask uses TCP to communicate between client, scheduler and workers by default. This means that all of these components must be on a TCP/IP network with open routes between the machines. Many connectivity problems step from firewalls or private networks blocking connections between certain components. An example of this would be running Dask on a cloud platform like AWS, but running the Python session and client on your laptop while sitting in a coffee shop using the free wifi. You must ensure you are able to route traffic between components, either by exposing the Dask cluster to the internet or by connecting your laptop to the private network via a VPN or tunnel.
```

> There is also [ongoing work to add support for UCX](https://blog.dask.org/2019/06/09/ucx-dgx) to Dask, which will allow it to make use of Infiniband or NVLINK networks where they are available.

Reviewer: ```suggestion
There is also [ongoing work to add support for UCX](https://blog.dask.org/2019/06/09/ucx-dgx) to Dask, which will allow it to make use of InfiniBand or NVLink networks where they are available.
```

This tool is targeted at IT administrators who want to enable their users to create Dask clusters, but want to maintain some centralized control instead of each user creating their own thing. This can also be useful for tracking Dask usage and setting per user limits.

```python
from dask.distributed import Client
from dask_gateway import GatewayCluster

cluster = GatewayCluster()
client = Client(cluster)
```

For each user the commands for creating and using a gateway cluster are the same. It is down to the administrator to setup and manage the gateway server and configure [authentication via kerberos or Jupyter Hub](https://gateway.dask.org/authentication.html#). They should also provide [configuration](https://gateway.dask.org/configuration-user.html) to their users so that Dask Gateway knows how to connect to the gateway server. In a large organisation or institution the IT department also likely provisions the machines that staff are using, and so should be able to drop configuration files onto users computers.

### Local CUDA Cluster

The last cluster manager I'm going to cover is `LocalCUDACluster` from the [Dask CUDA](https://github.com/rapidsai/dask-cuda) package.

> This is slightly different than the other cluster managers in that it is constructing a Dask cluster which is specifically optimised for a single piece of hardware. In this case it is targeting machines with multiple GPUs like an [NVIDIA DGX-2](https://www.nvidia.com/en-gb/data-center/dgx-2/).

Reviewer: Maybe here we should be more generic? Specifically pointing to a DGX-2 may give the impression users need a big machine, but they could be using that on any machine with at least one NVIDIA GPU, even a laptop.

Author: Good point!

Author: ```suggestion
This is slightly different than the other cluster managers in that it is constructing a Dask cluster which is specifically optimised for a single piece of hardware. In this case it is targeting machines with GPUs ranging from your laptop with an onboard NVIDIA GPU to an [NVIDIA DGX-2](https://www.nvidia.com/en-gb/data-center/dgx-2/) with multiple GPUs running in your datacentre.
```

The cluster manager closely follows the `LocalCluster` in that is creates resources locally on the current machine, but instead of creating one worker per CPU core it creates one per GPU. It also changes some of the configuration defaults to ensure good performance of GPU workloads.

```python
from dask.distributed import Client
from dask_cuda import LocalCUDACluster

cluster = LocalCUDACluster()
client = Client(cluster)
```

This package also has an alternative Dask worker bash command called `dask-cuda-worker` which also modified the defaults of the Dask worker to ensure it is optimised for GPU work.

## Future

Now that we have laid out the current state of the Dask distributed cluster ecosystem let's discuss where we could go next.

> As shown at the beginning a Dask cluster is a combination of scheduler, workers and clients which enable distributed execution of Python functions. Setting up your own cluster on your own machines is straight forward, but there is such a variety of ways to provision infrastructure that we now have a number of ways of automating this.

Reviewer: "scheduler, workers and clients", is clients really plural? I've never used multiple clients, maybe is just something I'm not aware of, just pointing this out to double-check whether this is a typo or is really something that can be done.

Author: It is possible to connect multiple clients. I've seen it done in some use cases, but it's probably not the primary use case. So changing this I think makes sense.

I'm actually working on adding clients to the cluster map dashboard, mainly just so that people can see when they've accidentally created the client object multiple times 😆.

Author: ```suggestion
As shown at the beginning a Dask cluster is a combination of scheduler, workers and client which enable distributed execution of Python functions. Setting up your own cluster on your own machines is straight forward, but there is such a variety of ways to provision infrastructure that we now have a number of ways of automating this.
```
